// apipulse-node/dist/config/AppConfig.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var AppConfig = /** @class */ (function () {
    function AppConfig() {
    }
    AppConfig.setConfig = function (options) {
        this.app = options.app;
        this.url = options.url;
        this.applicationName = options.applicationName;
        this.authKey = options.authKey;
        this.environment = options.environment;
        this.teamName = options.teamName;
        this.partnerId = options.partnerId;
    };
    AppConfig.getApp = function () {
        return this.app;
    };
    AppConfig.getUrl = function () {
        return this.url;
    };
    AppConfig.getApplicationName = function () {
        return this.applicationName;
    };
    AppConfig.getAuthKey = function () {
        return this.authKey;
    };
    AppConfig.getEnvironment = function () {
        return this.environment;
    };
    AppConfig.getTeamName = function () {
        return this.teamName;
    };
    AppConfig.getPartnerId = function () {
        return this.partnerId;
    };
    return AppConfig;
}());
exports.default = AppConfig;
