// apipulse-node/dist/buffer/ApiBufferKey.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ApiBufferKey = /** @class */ (function () {
    function ApiBufferKey(uri, method) {
        this.uri = uri;
        this.method = method;
    }
    ApiBufferKey.getApiBufferKeyFromObservedApi = function (observedApi) {
        try {
            return new ApiBufferKey(observedApi.getUri(), observedApi.getMethod());
        }
        catch (error) {
            console.error("Failed to create ApiBufferKey from ObservedApi:", error);
            throw error;
        }
    };
    ApiBufferKey.getApiBufferKeyFromApiConfig = function (apiConfig) {
        try {
            return new ApiBufferKey(apiConfig.getUri(), apiConfig.getMethod());
        }
        catch (error) {
            console.error("Failed to create ApiBufferKey from ApiConfig:", error);
            throw error;
        }
    };
    ApiBufferKey.prototype.equals = function (object) {
        if (!object) {
            return false;
        }
        if (object instanceof ApiBufferKey) {
            var apiBufferKey = object;
            return (this.uri.equals(apiBufferKey.uri) && this.method === apiBufferKey.method);
        }
        return false;
    };
    ApiBufferKey.prototype.getUri = function () {
        return this.uri;
    };
    ApiBufferKey.prototype.setUri = function (uri) {
        this.uri = uri;
    };
    ApiBufferKey.prototype.getMethod = function () {
        return this.method;
    };
    ApiBufferKey.prototype.setMethod = function (method) {
        this.method = method;
    };
    return ApiBufferKey;
}());
exports.default = ApiBufferKey;
