// apipulse-node/dist/common/constant/ApiMonitorCommonConstant.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Framework = exports.ConfigErrorCode = exports.ContentType = exports.HttpRequestMethod = exports.DataCategory = exports.DataSizeUnit = exports.ApiMonitorCommonConstant = void 0;
// Define a class to hold common constants
var ApiMonitorCommonConstant = /** @class */ (function () {
    function ApiMonitorCommonConstant() {
    }
    ApiMonitorCommonConstant.DATA_CATEGORY_MESSAGE_ATTRIBUTE_KEY = "DATA_CATEGORY";
    ApiMonitorCommonConstant.CONTENT_TYPE_HEADER_LOWER_CASE = "content-type";
    ApiMonitorCommonConstant.CONTENT_TYPE_HEADER_JSON_VALUE = "application/json";
    ApiMonitorCommonConstant.MASKING_RULES_CONFIGURATION_NAME = "PIIMaskingRules";
    ApiMonitorCommonConstant.DEFAULT_MASK_REPLACEMENT = "MASKED";
    ApiMonitorCommonConstant.MASKED_VALUE_SUFFIX = "SL_MASK";
    return ApiMonitorCommonConstant;
}());
exports.ApiMonitorCommonConstant = ApiMonitorCommonConstant;
// Enum for data size units
var DataSizeUnit;
(function (DataSizeUnit) {
    DataSizeUnit["KB"] = "KB";
    DataSizeUnit["MB"] = "MB";
    DataSizeUnit["GB"] = "GB";
})(DataSizeUnit = exports.DataSizeUnit || (exports.DataSizeUnit = {}));
// Enum for data categories
var DataCategory;
(function (DataCategory) {
    DataCategory["SAMPLE_DATA"] = "SAMPLE_DATA";
})(DataCategory = exports.DataCategory || (exports.DataCategory = {}));
// Enum for HTTP request methods
var HttpRequestMethod;
(function (HttpRequestMethod) {
    HttpRequestMethod["GET"] = "GET";
    HttpRequestMethod["HEAD"] = "HEAD";
    HttpRequestMethod["POST"] = "POST";
    HttpRequestMethod["PUT"] = "PUT";
    HttpRequestMethod["PATCH"] = "PATCH";
    HttpRequestMethod["DELETE"] = "DELETE";
    HttpRequestMethod["OPTIONS"] = "OPTIONS";
    HttpRequestMethod["TRACE"] = "TRACE";
})(HttpRequestMethod = exports.HttpRequestMethod || (exports.HttpRequestMethod = {}));
// Enum for content types
var ContentType;
(function (ContentType) {
    ContentType["JSON"] = "application/json";
    ContentType["TEXT"] = "text/plain";
    ContentType["UNKNOWN"] = "unknown";
})(ContentType = exports.ContentType || (exports.ContentType = {}));
// Enum for configuration error codes
var ConfigErrorCode;
(function (ConfigErrorCode) {
    ConfigErrorCode["TIMEOUT"] = "TIMEOUT";
    ConfigErrorCode["PARSE_ERROR"] = "PARSE_ERROR";
    ConfigErrorCode["INVALID_CONFIG"] = "INVALID_CONFIG";
})(ConfigErrorCode = exports.ConfigErrorCode || (exports.ConfigErrorCode = {}));
// Enum for frameworks
var Framework;
(function (Framework) {
    Framework["EXPRESS"] = "EXPRESS";
    Framework["HAPI"] = "HAPI";
})(Framework = exports.Framework || (exports.Framework = {}));
